### PR TITLE
GH-125837: Split `LOAD_CONST` into three.

### PR DESCRIPTION
**Pull request by markshannon**
Splits `LOAD_CONST` into three instructions
* `LOAD_INT` for ints in `range(256)`. Avoids the need for a space in the `co_consts` tuple and avoids an incref
* `LOAD_CONST_IMMORTAL` for other immortal objects. Avoids an incref
* `LOAD_CONST` for the remaining constants

* Issue: gh-125837

----
📚 Documentation preview 📚: https://cpython-previews--125972.org.readthedocs.build/

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -255,10 +255,28 @@ dummy_func(
...
+        inst(LOAD_CONST_IMMORTAL, (-- value)) {
+            PyObject *obj = GETITEM(FRAME_CO_CONSTS, oparg);
+            assert(_Py_IsImmortal(obj));
+            value = PyStackRef_FromPyObjectImmortal(obj);
+        }
+
+        replicate(4) inst(LOAD_INT, (-- value)) {
+            /* Tell code generator that this is a const load */
+            (void)FRAME_CO_CONSTS;
+            assert(oparg < 256);
```
Comment (picnixz) [inline Python/bytecodes.c]: Would it make sense to check for `(-_PY_NSMALLNEGINTS <= oparg && oparg < _PY_NSMALLPOSINTS)` instead?

File: Python/compile.c [hunk 2]
```diff
@@ -447,6 +447,8 @@ _PyCompile_DictAddObj(PyObject *dict, PyObject *o)
 Py_ssize_t
 _PyCompile_AddConst(compiler *c, PyObject *o)
 {
+
+
```
Comment (picnixz) [inline Python/compile.c]:
```suggestion
```

File: Python/flowgraph.c [hunk 3]
```diff
@@ -283,7 +283,7 @@ dump_instr(cfg_instr *i)
 static inline int
 basicblock_returns(const basicblock *b) {
     cfg_instr *last = basicblock_last_instr(b);
-    return last && (last->i_opcode == RETURN_VALUE || last->i_opcode == RETURN_CONST);
+    return last && last->i_opcode == RETURN_VALUE);
```
Comment (picnixz) [inline Python/flowgraph.c]:
```suggestion
    return last && last->i_opcode == RETURN_VALUE;
```

**Comment by picnixz** [comment]
> LOAD_INT for ints in range(256)

I think the range is -5 -> 256 (included) right? (at least _PyLong_SMALL_INTS works that way, but maybe 256 should be excluded even though it's stored in the _PyLong_SMALL_INTS).

File: Python/bytecodes.c (hunk 1 above)
Comment (markshannon) [reply Python/bytecodes.c]: `oparg` is unsigned, so it will always be `>= 0`.
I'll add an assert that `oparg < _PY_NSMALLPOSINTS`

**Comment by markshannon** [comment]
> I think the range is -5 -> 256 (included) right? (at least _PyLong_SMALL_INTS works that way, but maybe 256 should be excluded even though it's stored in the _PyLong_SMALL_INTS).

I want `LOAD_INT 0` to push `0` not `-5` to keep things simple.
So `LOAD_INT` will only cover `range(256)` even though the range of statically allocated ints is larger.

We can always add -5,-4,-3,-2 and -1 to `LOAD_COMMON_CONST` later if we want.

**Comment by markshannon** [comment]
[Performance for tier 1](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20241025-3.14.0a1+-f934559) appears to be about 0.5% faster.

[Stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20241025-3.14.0a1%2B-f934559/bm-20241025-azure-x86_64-faster%252dcpython-split_load_const-3.14.0a1%2B-f934559-pystats-vs-base.md) show the following fraction of constants loaded:

Opcode |  Fraction
--- | ---
LOAD_INT | 47%
LOAD_CONST_IMMORTAL | 42%
LOAD_CONST | 11%

With a reduction in increfs of immortal objects of about 20%.

File: Python/codegen.c [hunk 4]
```diff
@@ -280,6 +280,14 @@ codegen_addop_noarg(instr_sequence *seq, int opcode, location loc)
 static int
 codegen_addop_load_const(compiler *c, location loc, PyObject *o)
 {
+    if (PyLong_CheckExact(o)) {
+        int overflow;
+        long val = PyLong_AsLongAndOverflow(o, &overflow);
+        if (!overflow && val >= 0 && val < 256 && val < _PY_NSMALLPOSINTS) {
```
Comment (Eclips4) [inline Python/codegen.c:286]: Is `val < _PY_NSMALLPOSINTS` needed here?

Comment (Eclips4) [reply Python/codegen.c:286]: Oh, yes, it's matches the behavior of `bytecodes.c` (https://github.com/python/cpython/pull/125972#discussion_r1816437449)

File: Lib/test/test_ast/test_ast.py [hunk 5]
```diff
@@ -2303,15 +2303,14 @@ def get_load_const(self, tree):
...
-            if instr.opname == 'LOAD_CONST' or instr.opname == 'RETURN_CONST':
+            if instr.opname in ('LOAD_CONST', 'LOAD_CONST_IMMORTAL'):
                 consts.append(instr.argval)
         return consts
 
     @support.cpython_only
     def test_load_const(self):
         consts = [None,
                   True, False,
-                  124,
```
Comment (Eclips4) [inline Lib/test/test_ast/test_ast.py:2314]: Why was `124` removed? I think that we should back `124` here and just add `LOAD_INT` to tuple of possible bytecodes on line 2306

File: Lib/test/test_ast/test_ast.py [hunk 6]
```diff
@@ -2303,15 +2303,14 @@ def get_load_const(self, tree):
         co = compile(tree, '<string>', 'exec')
         consts = []
         for instr in dis.get_instructions(co):
-            if instr.opname == 'LOAD_CONST' or instr.opname == 'RETURN_CONST':
+            if instr.opname in ('LOAD_CONST', 'LOAD_CONST_IMMORTAL'):
```
Comment (Eclips4) [inline Lib/test/test_ast/test_ast.py]: Since `dis.get_instructions` is called without an `adaptive=True` arg, I think it's impossible to have a `LOAD_CONST_IMMORTAL` here?

File: Lib/test/test_ast/test_ast.py (hunk 5 above)
Comment (markshannon) [reply Lib/test/test_ast/test_ast.py:2314]: I've put it back.

File: Include/internal/pycore_magic_number.h [hunk 7]
```diff
@@ -261,6 +261,7 @@ Known values:
     Python 3.14a1 3606 (Specialize CALL_KW)
     Python 3.14a1 3607 (Add pseudo instructions JUMP_IF_TRUE/FALSE)
     Python 3.14a1 3608 (Add support for slices)
+    Python 3.14a2 3609 (Add LOAD_INT and LOAD_CONST_IMMORTAL instructions)
```
Comment (iritkatriel) [inline Include/internal/pycore_magic_number.h]:
```suggestion
    Python 3.14a2 3609 (Add LOAD_INT and LOAD_CONST_IMMORTAL, remove RETURN_CONST)
```

File: Lib/test/test_ast/test_ast.py [hunk 8]
```diff
@@ -2303,7 +2303,7 @@ def get_load_const(self, tree):
         co = compile(tree, '<string>', 'exec')
         consts = []
         for instr in dis.get_instructions(co):
-            if instr.opname == 'LOAD_CONST' or instr.opname == 'RETURN_CONST':
+            if instr.opname in ('LOAD_CONST', 'LOAD_CONST_IMMORTAL', 'LOAD_INT'):
```
Comment (iritkatriel) [inline Lib/test/test_ast/test_ast.py]: I think this should work:

```suggestion
            if dis.opmap[instr.opname] in dis.hasconst:
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-25-15-56-14.gh-issue-125837.KlCdgD.rst [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Adds :opcode:`LOAD_INT` and :opcode:`LOAD_CONST_IMMORTAL` instructions.
+``LOAD_INT`` pushes a small integer equal to the ``oparg`` to the stack.
+``LOAD_CONST_IMMORTAL`` does the same as ``LOAD_CONST`` but is more
+efficient for immortal objects.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-25-15-56-14.gh-issue-125837.KlCdgD.rst:4]: and removes `RETURN_CONST`.

File: Python/specialize.c [hunk 10]
```diff
@@ -465,6 +466,21 @@ _PyCode_Quicken(PyCodeObject *code)
             }
             i += caches;
         }
+        else if (opcode == LOAD_CONST) {
+            /* We can't do this in the bytecode compiler as
+             * marshalling can intern strings and make them immortal. */
+            oparg = (oparg << 8) | instructions[i].op.arg;
```
Comment (iritkatriel) [inline Python/specialize.c]: I'd put this above, before "if (caches)". Then below you don't need the same line again in case of `EXTENDED_ARG`, just `if (opcode != EXTENDED_ARG) oparg = 0;`.

This will also future proof this in case another case is added that needs the `oparg`.

File: Python/flowgraph.c [hunk 11]
```diff
@@ -2436,7 +2419,7 @@ convert_pseudo_conditional_jumps(cfg_builder *g)
 }
 
 static int
-convert_pseudo_ops(cfg_builder *g)
+convert_pseudo_ops(cfg_builder *g, int is_generator)
```
Comment (iritkatriel) [inline Python/flowgraph.c]: `is_generator` is not used.

File: Python/codegen.c [hunk 12]
```diff
@@ -656,6 +664,9 @@ codegen_setup_annotations_scope(compiler *c, location loc,
         codegen_enter_scope(c, name, COMPILE_SCOPE_ANNOTATIONS,
                             key, loc.lineno, NULL, &umd));
 
+    // Insert None into consts to prevent an annotation
+    // apperaing to be a docstring
```
Comment (JelleZijlstra) [inline Python/codegen.c]:
```suggestion
    // appearing to be a docstring
```

File: Doc/library/dis.rst [hunk 13]
```diff
@@ -1086,6 +1079,22 @@ iterations of the loop.
    Pushes ``co_consts[consti]`` onto the stack.
 
 
+.. opcode:: LOAD_INT (i)
+
+   Pushes the integer ``i`` onto the stack.
+   ``i`` must be in ``range(256)``
```
Comment (iritkatriel) [inline Doc/library/dis.rst:1085]: Why the restriction on the range of `i`? Can't use `EXTENDED_ARG` with this bytecode?

Comment (markshannon) [reply Doc/library/dis.rst:1085]: We want to rapidly return a pre-allocated `int`, which is currently -5 to 257 (inclusive), as that is much faster than calling `PyLong_FromLong` or other API function.

Rather than worry about 257, is is simpler to just limit it to unextended instructions.

Comment (iritkatriel) [reply Doc/library/dis.rst:1085]: Can we call it `LOAD_SMALL_INT`?

Comment (markshannon) [reply Doc/library/dis.rst:1085]: We can. I have.

File: Python/codegen.c [hunk 14]
```diff
@@ -656,6 +664,9 @@ codegen_setup_annotations_scope(compiler *c, location loc,
         codegen_enter_scope(c, name, COMPILE_SCOPE_ANNOTATIONS,
                             key, loc.lineno, NULL, &umd));
 
+    // Insert None into consts to prevent an annotation
+    // appearing to be a docstring
+    _PyCompile_AddConst(c, Py_None);
```
Comment (iritkatriel) [inline Python/codegen.c:669]: What does this have to do with this PR?

File: Python/flowgraph.c [hunk 15]
```diff
@@ -2173,7 +2155,8 @@ remove_unused_consts(basicblock *entryblock, PyObject *consts)
 
     for (basicblock *b = entryblock; b != NULL; b = b->b_next) {
         for (int i = 0; i < b->b_iused; i++) {
-            if (OPCODE_HAS_CONST(b->b_instr[i].i_opcode)) {
+            int opcode = b->b_instr[i].i_opcode;
+            if (OPCODE_HAS_CONST(opcode) && opcode != LOAD_SMALL_INT) {
```
Comment (iritkatriel) [inline Python/flowgraph.c]: `hasconst` is defined in the `dis` docs as "Sequence of bytecodes that access a constant". I think it will make more sense if we clarify that it means "access a constant from `co_consts`" and then remove `LOAD_SMALL_INT` from `hasconst` and `OPCODE_HAS_CONST`.

File: Python/codegen.c (hunk 14 above)
Comment (markshannon) [reply Python/codegen.c:669]: The consts for an `__annotate__` function could be `(0, "a_string")`. Removing the `0` makes "a_string" the doc string.

https://github.com/python/cpython/issues/126072

File: Python/flowgraph.c (hunk 15 above)
Comment (markshannon) [reply Python/flowgraph.c]: That makes sense. `LOAD_COMMON_CONST` is not in `hasconst`.

File: Lib/test/test_ast/test_ast.py [hunk 16]
```diff
@@ -2303,15 +2303,14 @@ def get_load_const(self, tree):
...
-            if instr.opname == 'LOAD_CONST' or instr.opname == 'RETURN_CONST':
+            if instr.opcode in dis.hasconst:
                 consts.append(instr.argval)
         return consts
 
     @support.cpython_only
     def test_load_const(self):
         consts = [None,
                   True, False,
-                  124,
```
Comment (Eclips4) [inline Lib/test/test_ast/test_ast.py:2314]: Can we replace it with an int which is not captured by `LOAD_SMALL_INT` instead of removing it? For example, 1000.

File: Python/flowgraph.c (hunk 15 above)
Comment (Eclips4) [reply Python/flowgraph.c]: Maybe it's worth testing whether creating integers within the `range(256)` will not affect `co_consts`?

File: Lib/test/test_ast/test_ast.py (hunk 16 above)
Comment (markshannon) [reply Lib/test/test_ast/test_ast.py:2314]: done

File: Python/flowgraph.c (hunk 15 above)
Comment (markshannon) [reply Python/flowgraph.c]: I don't want to go overboard on testing internal details of the bytecode compiler.
`LOAD_CONST` with a small `int` is still correct, just a bit less efficient than `LOAD_SMALL_INT`.
We already have plenty of tests in `test_dis` that test this

**Review by iritkatriel — APPROVED** [review]
